### DISCUSSION TITLE
Flight Management

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#25] [Ideas]
We need some way to keep track of individual sessions. In Thorium Classic, these are called "flights". Typically a single flight runs for the duration of a mission, and is then deleted (although flights can be paused, saved, and reloaded later).

Should Thorium Nova be any different? You start a flight, play through a mission, and then you're done. That's that.

A few reasons why we might want to take a different approach:

- Campaign missions, where the same crew flies the same ship over the course of several missions.
- Persistent universe, where the actions that the crew takes affects how future missions work. 
  - As an aside, I've been playing Skyrim for the first time recently, and it boggles my mind that some quests end in the death of an NPC, and they're just gone. Forever. Kind of weird.
- Add an RPG aspect to the game, where you can upgrade your ship mid- or post-mission with better engines, weapons, armor, etc. as a means of leveling up.

There might be other reasons I might be missing. A few reasons against:

- Many of the things above work well for single player games, but for cooperative multiplayer games, it's tough to get the gang together over and over again to continue playing through their game (although, theoretically, with a good enough experience, people will really _want_ to come back and play Thorium Nova again.
- It adds complexity to the project which might not be necessary or desired. Especially figuring out how events of one mission affect future missions. How do you even model that?
- There's nothing stopping campaign missions from happening without a persistent universe - you just start from square one at the beginning of every mission. This is how the Space EdVenture centers in Utah do campaign missions.

Anyone have any other thoughts? My default is to go with a non-persistent universe, since its easier and not having a persistent universe doesn't limit the ability to implement other features. I'd love everyone else's thoughts though.

**Comment by Xansta** [comment]
The default should be the non-persistent universe. However, the engine should support a persistent universe if a mission writer wants to either link several missions together or have a flight stop, then resume. I've often had the experience of writing a mission and my users/players/crew not being able to finish it because they have other commitments in real life. The mission I thought might take 30 minutes goes on for three hours and only 25% of it is complete. It's not like there's nothing to do, but they have not solved a part of the puzzles, have not reached a critical milestone or are distracted by wanting to shoot all the enemies they come across instead of pursuing the mission goals. 

Btw, I tend to state things firmly. My ideas are intended to be conversation openers, or invitations to brainstorm.

**Reply by alexanderson1993** [reply]
Here's my thinking on this:

- Flights can be paused, saved to disk, and loaded at a later time.
- It should be possible to change the mission mid-way through a flight. 
  - This adds a bit of complexity, since missions will likely define a number of initialization steps at the beginning of the flight which might include placing the crew ship at the appropriate point in space, but you don't want the crew to randomly transport somewhere else in the middle of a flight. Or if one mission uses a totally different star map than another mission. These don't make it impossible to do this - they're just wrinkles that need to be ironed out.

In other words, I think we can do what you suggest, especially helping crews that only get part-way through a mission come back and finish the remainder.

**Reply by astrolamb-gaming** [reply]
When you say "change the mission mid-flight", I'm not sure what that would realistically look like. 

I would propose more of an _objective_ based system. Given a certain objective, a series of tasks (i guess it's kinda like quests in an rpg) need accomplished, sometimes in a particular order. 

But if desired, a crew could change their main objective to something else. 

This gets a bit into how travel will happen. If an objective changes and it is far away, how do the players get there? We've discussed that we want the players to have a choice in what mode of long distance travel they want to use, but we could also include a "fast travel" option.

**Reply by alexanderson1993** [reply]
I think all of these topics deserve discussions of their own, but we can start preliminary discussion here. Also, all of this is based on my own experience with Thorium Classic.

Before I talk about changing the mission mid-flight, let me describe what I think a mission is. Here's the [definition of Timeline in the glossary](https://github.com/Thorium-Sim/thorium-nova/blob/develop/GLOSSARY.md), which I use synonymous with mission:

> **Timeline**: A mission definition. The mission is the storyline; the timeline is how that storyline is codified into Thorium Nova. The timeline is configured as a series of steps, each step describing what the Flight Director should do as well as a number of Inputs (changes to the simulation) that should be triggered at that point in the mission. During the flight, the timeline is available on the Core for the Flight Director to review and advance.

Not noted in the glossary, but missions should also include a mission briefing of some kind, so the crew actually knows what they're doing and why they're doing it.

To expand on the timeline, it includes any initialization steps (creating the necessary entities, setting up event listeners which trigger inputs when the crew performs some action [entering an area of space, destroying a target, sending a message, etc], defining the initial mission objectives), mission content (long range messages, short range communications, special sensor scan results, etc), and the timeline of inputs that happen during a mission. 

For a Flight Director-run mission, the timeline would be manually advanced by the Flight Director at their whim. For a non-Flight Director mission, different inputs during the mission would have to be triggered manually by the crew performing some action, and those triggers would have to be set up at some point, either in the initialization step or by another trigger. 

My point is that missions are far more than just "Here's an objective, go do it." There's a lot of other stuff involved around getting them to that objective. Here's some of the inputs which are triggered by an actual Thorium Classic timeline:

- Change what's on the viewscreen (likely not relevant to Thorium Nova, since the viewscreen should be fully automated)
- Change the listed mission objectives
- Send messages to the bridge from someone on the ship
- Send a long range message to the crew from another ship
- Create sensor contacts and targeting contacts
- Send special sensors data
- Start a short-range communication signal.
- Damage or repair a ship system
- Play special sound effects
- Change the lighting to a specific setting

Maybe many of these things will just be automated as part of Thorium Nova, but I think we also need to give mission writers as many tools as we can to let them tell the story. This ties into https://github.com/Thorium-Sim/thorium-nova/discussions/28#discussioncomment-1444503 by letting mission writers give each member of the crew a way to tell the story through their stations rather than just have the Flight Director dictate the story over the microphone. (I have lots I could say about that too, but this is already getting long)

So, changing a mission mid-flight would involve loading the mission configuration into the current flight and run the initialization steps. I imagine both missions would run parallel, since entities for both would exist in the universe, and the event listeners for the first mission would still be active (although, for Flight Director run missions, I suppose it could be possible for them to choose to disable the event listeners for the previous mission).

That said, the scenario of starting a new mission in the middle of an existing mission seems unlikely. What's more likely is that the crew gets to the end of their current mission, and either some event listener triggers the new mission to start (creating a chain of missions, one after the other) or the Flight Director manually starts a new mission.

Or maybe a clever mission writer creates a hub of sorts where the crew can go at the end of their mission. There, they can find various NPCs who will offer them quests that start them off on new missions. I think that should be a possibility too.

> but we could also include a "fast travel" option.

I don't dislike this idea, but we would need to find a way to do this as part of multi-bridge flights where every participating crew is aware of fast traveling and consents. The way I see it, fast travel (or speeding up the passage of time, whatever you want to call it) only works if a) both ships execute fast travel at the same time, b) one ship can fast travel without affecting the passage of time for another ship (some kind of wormhole or instant transport without being OP) or c) fast travel is just allowed when there are more than 1 player ship in the flight.

Two possibilities: 

- Write some kind of wormhole generator device into the canon that allows instant one-way movement back to a base of some kind. Have all of the mission start at that location. I'm not a fan of this, but it solves the fast travel and multi-bridge issues at the same time.

- Have some kind of technology or ship system that allows fast travel. We did this at the Utah space center, where one ship had a Transwarp drive and docking ports which allowed the other ships at the space center (there were 5 bridges at the time) to dock with that ship, and then all of them jumped to transwarp together.

**Reply by astrolamb-gaming** [reply]
I think our perspectives on what missions would look like are actually quite aligned; commenting via phone probably contributes to shorter and therefore less accurate portrayals. I did not mean to convey that a mission is a one step process. 

This may be merely a semantics thing, but I guess where i differentiate from what you've stated is mission vs timeline. 

The _mission_, as i see it, is the process by which an objective is either achieved or failed. The scope of the objective, as well as all actions (including sub-objectives) needed to achieve it, are part and parcel of the mission. 

A _timeline_, on the other hand, is a series of steps (triggers or whatever)  that must be completed in order. 

I see a mission as a set of timelines, not just one, which could be arranged either in series or in parallel depending on the situation. I suppose you could even have a timeline within a timeline. 

Thinking as a gm, i would want the capability to trigger the start of a new timeline (either pre-defined or specified on the spot) at any point in a mission. 

Getting into plugins a bit, i think it'd be great to be able to define timelines and then publish them to be available for any gm. Depending on how these are defined, they would not need any special code, and so could fit under a content pack designation instead.

**Reply by astrolamb-gaming** [reply]
> That said, the scenario of starting a new mission in the middle of an existing mission seems unlikely. What's more likely is that the crew gets to the end of their current mission, and either some event listener triggers the new mission to start (creating a chain of missions, one after the other) or the Flight Director manually starts a new mission.

Yeah i agree that hopping between missions would be unlikely. 

> Or maybe a clever mission writer creates a hub of sorts where the crew can go at the end of their mission. There, they can find various NPCs who will offer them quests that start them off on new missions. I think that should be a possibility too.

Wholeheartedly agree with this

**Reply by alexanderson1993** [reply]
I agree with your distinction between mission and timeline - and that distinction should probably be built into the controls as well. Thorium Classic treats timelines and missions as the same thing, but you're right that they're not. As you say, Missions are composed of one or more timeline, either in series or parallel.

Series timelines should be familiar - one arc of the mission is completed, which triggers the next timeline to commence.

Parallel timelines are like overarching steps that happen throughout several arcs of the mission, or an internal second/B storyline, like security dealing with a mess hall fight, which may or may not be directly tied to the main storyline, and that runs independently of said main storyline.

In other words, we are indeed aligned.

**Comment by astrolamb-gaming** [comment]
Going back to the  OP (we got a bit sidetracked...):

> Campaign missions, where the same crew flies the same ship over the course of several missions.

I'm a big fan of this idea. I will note that the missions would not have to be specifically linked. there are some other aspects we could add. For example, say your ship takes some damage at one point and the crew can't completely repair it without spending a lot of time in drydock - the efficiency of some of the ship's systems could be reduced until they're able to return for repairs.  This would fit particularly well with the RPG idea.

> Persistent universe, where the actions that the crew takes affects how future missions work.

While I would love to see this implemented, I think that the vast majority of players who would want this will also have a Flight Director who can handle any changes needed. But I would fully support a 2.0 version which does include this.

> Add an RPG aspect to the game, where you can upgrade your ship mid- or post-mission with better engines, weapons, armor, etc. as a means of leveling up.

If you've ever played Pulsar: Lost Colony (an *FPS* bridge sim), it has an overaching story, and you can buy or pirate or just find new ship components (weapons, engines, etc) to upgrade your ship. 

I'd love to see a variant of this (obviously not FPS (though tbh I think a "landing party" mode could be cool), nor do I think an overarching story is necessary), but having a quantity of random events/encounters that can happen to a crew who are wandering the galaxy would be a great addition.

I imagine that racking changes to a vessel is a lot simpler than tracking every entity in the universe, so I deem that as a quite reasonable compromise.

**Reply by alexanderson1993** [reply]
> I imagine that racking changes to a vessel is a lot simpler than tracking every entity in the universe, so I deem that as a quite reasonable compromise.

Yeah, I'm not sure how this is going to work out entirely. I imagine that solar systems will have some automation to control automatically spawning random ships so it doesn't feel like a ghost town, and then missions will be able to spawn the specific entities that are needed for the story. Plus, we can set up the game loop to recognize which system the player ships are in, and to run the simulation less frequently for ships that are far away from those player ships.

**Reply by Xansta** [reply]
I like the distinction between timelines and missions. Allow me to give an example from a scenario I wrote for Empty Epsilon, Delta Quadrant Patrol Duty (spoilers follow for anyone that has not played this mission). The crew (or crews) are given orders to patrol between three friendly stations. The scenario tracks each time they dock at a station and tells them the least patrolled station (the one they should go to next) and the most patrolled station (the one they should not pick next, but instead choose one of the others). This fits nicely in the timeline paradigm. 

Another part of the scenario generates enemies that attack one of the patrol stations, the player or some of the other stations in the area. These stations are friendly, neutral or enemy stations. The criteria for these enemies to be generated range from duration of the game session so far to tying in to how far along their assigned patrol duties the players have reached. This makes the enemy spawning part timeline based and part independent.

There are several optional objectives available. Some are triggered by the progress made by the players on the patrol part of the scenario. Others are the result of scanning interesting artifacts placed either along the direct patrol routes or slightly off the beaten path. The optional missions have some kind of reward at the end of them typically involving some kind of improvement to their ship they can get when they dock at one of their patrol points. These include things like a 50% cooling improvement for the beam weapons, the addition of a missile tube (which later turns out to have issues and thus fluctuates between usable and broken), a 20% increase in the top speed of the impulse engines, a 100% increase in the ship's maneuverability, a 10% improvement on beam range, damage and recharge time. From these kinds of rewards, you can tell that the game has quite a focus on combat systems and better ways to (in this case) protect from marauding enemies. 

Each mission writer can decide what kind of rewards to grant. They could simply be additional story elements. 

A couple of the optional missions have timeline-esque properties where another part of the branch is triggered after certain circumstances are met.

I've got a couple of ambush events governed by not only how far they have progressed through the patrol mission, but their proximity to the nebula where the enemies are just waiting for them to come by.

Ultimately, once the patrol is complete, depending on the options selected at the start of the mission, the ship(s) are called to repel a large enemy invasion around one of the three patrol point stations. This is designed to be the grand finale. This is also where completion of some of the optional missions could facilitate completion of the final objective.

The first couple of scenarios I wrote were very much linear. The crew was expected to follow the story line based on the information given and the points of the story rolled out in a linear fashion. Delta Quadrant Patrol Duty was one that I wrote a bit later where I wanted it to be more open and unpredictable in what would happen each time you played the scenario. There are aspects that the crew might miss entirely depending on what they decide to do. For these missions, I had to be sure that the crew got the message clearly as to what their next goal was or that they had a way to check on what they were supposed to do. They were specifically designed to run to completion without any GM intervention. Obviously, communicating timeline or mission goals can be handled by the GM when the players don't quite get what's going on or when they miss something.

I want to be sure your timeline/mission/scenario engine supports linear stories as well as more objective based stories (if you were wondering why the long explanation). 

Your mission writers will likely fall into two camps: those that expect the GM to be present to help guide the participants along the story and those that expect no GM to be present. I've got experience with both sets of assumptions. I like writing open ended non-GM scenarios with semi-undefined outcomes because I get to participate on the crew during those scenarios/missions/timelines and not have complete foreknowledge of what's going to happen next. @astrolamb-gaming will tell you I also like to GM sessions with players to give them something different from any stock missions/scenarios/timelines that come with the game.

**Reply by alexanderson1993** [reply]
> I want to be sure your timeline/mission/scenario engine supports linear stories as well as more objective based stories (if you were wondering why the long explanation).

I don't see how these two goals are in conflict. Both should be possible.

All of this context is really helpful for helping me get a mental model for how timelines should work in Thorium Nova, especially compared to Thorium Classic. Thus far, we'll need:

- The ability to set up event listeners which trigger other mission actions
- Variables which can be referenced in a later timeline step or event listener, to track whether the crew has completed some task or done some thing
- Conditional triggers for whether an event or action should be triggered or not.
- A few utilities for mission writers, like rolling a dice.

My hope is that the UI can be similar to iOS shortcuts, which is basically a simple programming environment which still has a _lot_ of power.

**Reply by alexanderson1993** [reply]
> Your mission writers will likely fall into two camps: those that expect the GM to be present to help guide the participants along the story and those that expect no GM to be present. 

I'm afraid this likely will be the case. Some missions will only be written for crews with GMs, although I imaging we could make it so missions that could be run without a GM have the ability for a GM to step in, if desired.

**Reply by astrolamb-gaming** [reply]
Do you have a plan for how we will define timelines/missions? Will it be some sort of list of commands? Text based or otherwise?

**Reply by alexanderson1993** [reply]
That's a discussion all on its own! 😆

I'll have to think about this, or someone else will have to come up with a good proposal. The separation of "missions" and "timelines" has got me thrown for a loop.

Also, timelines might be a poor paradigm for non-GM lead missions, since it isn't exactly a linear sequence of steps, but rather a series of event listeners that could be triggered at any moment, triggering actions and setting up more event listeners ad infinitum. I don't have a good way to visualize that yet.

As for text-based vs. Thorium Nova UI only, that's also up for discussion. Text based, with a solid UI on top of it, seems a good idea so long as the text format isn't too complicated.

**Reply by alexanderson1993** [reply]
That said (about text-based), if we can make the editing experience so good that any use-case you would want a text-based editor for (copy/paste, selecting large chunks to rearrange, etc) is covered in the Thorium Nova UI, then maybe we just do the UI.

**Reply by alexanderson1993** [reply]
Okay, I've never actually looked at an EE mission script before. I didn't realize it was written in Lua.

https://github.com/daid/EmptyEpsilon/blob/master/scripts/scenario_01_quick_basic.lua

I mean, using a programming language to write a mission is pretty awesome, not gonna lie. But I also fear that might discourage non-technical people from writing missions. Then again, if this is a good incentive to teach them a programming language, that seems like a bonus to me. How many of us learned programming by hacking on our favorite games?

This might also fill the "mod" niche. Custom behaviors could be added to things by adding event listeners and then responding to those events by executing some custom code. To use an example from another discussion, a "torpedo detonated" event would provide the event listener with the last position of the torpedo, which could then be used to spawn a black hole. 

Granted, I think we could create a UI that could do the same thing without needing to write programming language, no-code style.

This right here deserves it's own discussion too.

**Reply by astrolamb-gaming** [reply]
> Also, timelines might be a poor paradigm for non-GM lead missions, since it isn't exactly a linear sequence of steps, but rather a series of event listeners that could be triggered at any moment, triggering actions and setting up more event listeners ad infinitum. I don't have a good way to visualize that yet.

I visualize timelines as quests like in skyrim if that helps. I have thoughts for that but i think that would be more suited to a live discussion than text

**Reply by alexanderson1993** [reply]
I might have to write a proposal for another live meeting sometime soon...

Also, thank you for this evidence that I can give my wife that playing Skyrim is, indeed, product research 😉

**Comment by Ryan-Anderson-PhD** [comment]
Our model of flying would really benefit from the ability to carry things over from flight to flight in a "campaign" style mode.

**Reply by Arcadianer** [reply]
Same here. Lately I have been playing D&D with a few friends in a VTT called Foundry. They have an interesting system where each Game piece on the Virtual Battlefield is also linked to a Character in the background. If something happens to the Game piece on the field, it changes the Character in the Database.

So my proposal is to do something hybrid between Classic and an RPG.
Have the World, the Scenario and the playable pieces separated.

- The World defines all Systems, the Planets, the automated Patrols that make the world feel alive.

- The Scenario defines the Mission, the Objectives, the NPCs for this Adventure and how they all interact

- The Playable pieces are the ships that the Players fly or some important NPC. Their health, their hardware and their configuration are synced and saved persistent, so a ship can be used in different Scenarios.

**Reply by alexanderson1993** [reply]
> The Playable pieces are the ships that the Players fly ... so a ship can be used in different Scenarios.

I like this. Basically saving a ship after a flight so it can be used as the player ship in a new, future flight.

Otherwise, I think my vision for flights and missions covers the World, Scenario, Players model that you describe.

- The World is defined by the plugins which are loaded for a flight, which have the systems, stars, and planets.
- The Scenario is the mission/timeline that you pick for a given flight.
- The Player is the ship which you choose for the player at the beginning of the flight.

**Comment by Ryan-Anderson-PhD** [comment]
This sounds like a good model to me. It allows for a living universe, but
for scriptable events and continuity of some pieces.

**Comment by alexanderson1993** [comment]
Just so everyone knows, this is the very next thing that I'm going to be working on. I'm going to create the issues in Github for this sometime this evening or tomorrow.
